docs/manpages/tracker-import.1
.TH tracker-import 1 "July 2009" GNU "User Commands"

.SH NAME
tracker-import \- Imports Turtle file data into the database

.SH SYNOPSIS
\fBtracker-import\fR 
\fIFILE\fR

.SH DESCRIPTION
.B tracker-import
allows data to be imported to the database by providing files with
Turtle content.

Multiple \fIFILE\fR arguments can be provided to import data from
multiple files.

The \fIFILE\fR argument can be either a local path or a URI. It also
does not have to be an absolute path.

.SH OPTIONS
.TP
.B \-?, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Print version.

.SH SEE ALSO
.BR tracker-store (1),
.BR tracker-sparql (1),
.BR tracker-info (1).
.TP
.BR Turtle.


NEWS
NEW in 0.7.8
============

Improvements / New:

  General:
  * Add --version option to command line tools.

  Ontology:
  * NCO: Add more IM properties.
  * NMO: Add first draft of SMS support.
  * Work on ontology description and documentation.

  SPARQL:
  * Limit scope of blank nodes to conform to specification.
  * Fix missing results due to OPTIONAL optimization.

  libtracker-data:
  * Fix critical warning on tracker-store startup.

  tracker-extract:
  * Fix use of tagreadbin in GStreamer extractor.
  * Use NAO tags for image keywords.
  * Correctly detect 3GPP audio files in GStreamer extractor.
  * Extract nmm:albumTrackCount from MP3 files.

  RSS Reader:
  * Add some Zeitgeist interaction to the RSS Reader example.

Notes:

  The Deskbar applet packaged with Tracker is currently out of date
  and needs some love to get it working with the new APIs. It is in
  the same state as it was for 0.6. As such it isn't expected to work.


NEW in 0.7.7
============

Improvements / New:

  Ontology:
  * Mark nco:phoneNumber, nco:imID, and nco:emailAddress as inverse
    functional properties.

  SPARQL:
  * Optimize simple optionals with inverse functional properties.
  * Return property URI instead of internal uri for fts:offsets().

  libtracker-data:
  * Propagate SQLite errors on updates.

  tracker-miner-fs:
  * Handle errors when moving items recursively.


NEW in 0.7.6
============

Improvements / New:

  General:
  * Fixed libvorbis detection in configure.
  * Fixed data/.gitignore to not exclude .desktop.in files.
  * Added AC_CACHE_CHECK for ioprio in configure for cross-platform
    compilation.
  * Imported data generation scripts from old repository.

  Ontology:
  * Fixed nmm:exposureTime to be a double not integer.
  * Added tracker:weight to ontology properties (was only used on few
    properties before like nie:title). Used for scoring/ranking.
  * Added more FTS fields so FTS matching occurs on more useful data.
  * Added nco:metacontact concept
    (see http://live.gnome.org/Tracker/Documentation/ContactsOntology)
  * Removed cardinality restriction for dc:contributor.
  * Removed cardinality maximum from nco:contrbutor & nco:creator.
  * Renamed nmo:VOIPCall to nmo:Call.

  SPARQL:
  * Fixed fts:rank when fts:match is not first predicate in query.

  Functional Tests:
  * Added for fts:rank
  * Added for nco:metacontact

  libtracker-common:
  * Fixed tracker_coalesce() to not use empty/blank strings.
  * Fixed tracker_is_blank_string() to be UTF-8 compliant.
  * Added tracker_sparql_builder_object_double() method.

  libtracker-data:
  * Fixed assertion using cursors open twice (tracker-store crash).
  * Fixed outer query cursors being used by inner queries.

  libtracker-miner:
  * Fixed calls to tracker-store to be completely async (no blocking
    when tracker-store is not available)
  * Fixed trying to free invalid GPtrArray pointers.
  * Updated reference documentation.
  * Improve error reporting when initializing volumes.

  tracker-store:

  tracker-extract:
  * Fixed return value in albumart_set().
  * Fixed JPEG to coalesce nie:comment.
  * Fixed JPEG leak if file size is < 18 bytes.
  * Fixed MP3 id3v20/v22 memory corruption.
  * Fixed MP3 id3v20/v22 comments retrieval.
  * Fixed cases where _object_unvalidated() was used over _object().
  * Added album art support with libquill (alternative to GdkPixbuf).
  * Updated ALL plugins to use TrackerSparqlBuilder API (from older
    variant), this produces less D-Bus traffic and the output is
    generally nicer.

  tracker-search-bar:
  * Fixed to show all documents, not just paginated ones.
  * Fixed icon background to be transparent.
  * Fixed showing results and no results label at the same time.
  * Fixed window height to be a lot longer by default.
  * Fixed selecting blank rows for separators.
  * Fixed showing blank rows after all results.
  * Fixed sorting (broken by use of hash table).
  * Added vertical orientation for applet.
  * Added ranked sorting for searches.
  * Added limit for each category to 5.
  * Added interactive searching on a 300ms delay.

  tracker-search-tool:
  * Added interactive searching.

Bugs:

  * Fixes GB#600081, Get rid of Bonobo

Notes:

  The Deskbar applet packaged with Tracker is currently out of date
  and needs some love to get it working with the new APIs. It is in
  the same state as it was for 0.6. As such it isn't expected to work.


NEW in 0.7.5
============

Improvements / New:

  SPARQL:
  * Fixed to flush update buffer after DROP GRAPH.
  * Fixed leaking variables in reference cycle.
  * Support fts:rank and fts:offsets in function queries.

  libtracker-common:
  * Fixed battery handling so we don't throttle indexing when on low
    battery in cases where we are charging on AC power.
  * Fixed logging so that verbosity=1 displays all G_LOG_LEVEL_INFO
    messages. Previously it was unused.
  * Fixed vapi dependencies in Makefile.am
  * Added tracker_info() and tracker_dbus_request_info() to use
    G_LOG_LEVEL_INFO logging.

  libtracker-db:
  * Added tracker_db_statement_bind_null(). This allows use of ?
    parameters in SQL to be set to NULL. No longer special casing NULL
    when used as a value when deleting single valued property values.

  libtracker-data:
  * Fixed crash from dereferencing GError when NULL in
    delete_metadata_decomposed(). 
  * Added a rollback callback to prevent sending update events which
    are rolled back.

  libtracker-fts:
  * Use tracker:weight for per-property weight.

  libtracker-miner:
  * Fixed overwriting throttle in TrackerMinerFiles class
    construction.
  * Fixed TrackerCrawler class vtable definition of ::finished.
  * Fixed g_critical() if .service files are not available for some
    miners (this is the case for the Evolution plugin which is not
    started by D-Bus but Evolution).
  * Fixed tracker_connect() call to disable D-Bus timeouts.

  tracker-store:

src/tracker-utils/tracker-import.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/*
 * Copyright (C) 2006, Mr Jamie McCracken (jamiemcc@gnome.org)
 * Copyright (C) 2008-2009, Nokia
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <locale.h>
#include <stdint.h>

#include <glib.h>
#include <glib/gi18n.h>

#include <libtracker-client/tracker.h>
#include <libtracker-common/tracker-common.h>

#define ABOUT                                                             \
        "Tracker " PACKAGE_VERSION "\n"

#define LICENSE                                                           \
        "This program is free software and comes without any warranty.\n" \
        "It is licensed under version 2 or later of the General Public "  \
        "License which can be viewed at:\n"                               \
        "\n"                                                              \
        "  http://www.gnu.org/licenses/gpl.txt\n"


static gchar        **filenames = NULL;
static gboolean       print_version;

static GOptionEntry   entries[] = {
	{ "version", 'V', 0, G_OPTION_ARG_NONE, &print_version,
	  N_("Print version"),
	  NULL,
	},
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames,
	  N_("FILE"),
	  N_("FILE")},
	{ NULL }
};

int
main (int argc, char **argv)
{
	TrackerClient	*client;
	GOptionContext	*context;
	gchar          **p;

	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	/* Translators: this messagge will apper immediately after the	*/
	/* usage string - Usage: COMMAND [OPTION]... <THIS_MESSAGE>	*/
	context = g_option_context_new (_("- Import data using Turtle files"));

	/* Translators: this message will appear after the usage string */
	/* and before the list of options.				*/
	g_option_context_add_main_entries (context, entries, NULL);
	g_option_context_parse (context, &argc, &argv, NULL);

	if (print_version) {
                g_print ("\n" ABOUT "\n" LICENSE "\n");
		g_option_context_free (context);

		return EXIT_SUCCESS;
	}

	if (!filenames) {
		gchar *help;

		g_printerr ("%s\n\n",
			    _("One or more files have not been specified"));

		help = g_option_context_get_help (context, TRUE, NULL);
		g_option_context_free (context);
		g_printerr ("%s", help);
		g_free (help);

		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	client = tracker_connect (FALSE, G_MAXINT);

	if (!client) {
		g_printerr ("%s\n",
			    _("Could not establish a D-Bus connection to Tracker"));
		return EXIT_FAILURE;
	}

	for (p = filenames; *p; p++) {
		GError *error = NULL;
		GFile  *file;
		gchar  *uri;

		g_print ("%s:'%s'\n",
			 _("Importing Turtle file"),
			 *p);

		file = g_file_new_for_commandline_arg (*p);
		uri = g_file_get_uri (file);
		
		tracker_resources_load (client, uri, &error);
		
		g_object_unref (file);
		g_free (uri);
		
		if (error) {
			g_printerr ("%s, %s\n",
				    _("Unable to import Turtle file"),
				    error->message);
			
			g_error_free (error);
			continue;
		}

		g_print ("  %s\n", _("Done"));
		g_print ("\n");
	}

	tracker_disconnect (client);
	
	return EXIT_SUCCESS;
}
